pkg/api/query/grpc.go

```diff
@@ -202,24 +211,32 @@ func (g *GRPCAPI) QueryRange(request *querypb.QueryRangeRequest, srv querypb.Que
 	endTime := time.Unix(request.EndTimeSeconds, 0)
 	interval := time.Duration(request.IntervalSeconds) * time.Second
 
-	var engine promql.QueryEngine
 	engineParam := request.Engine
 	if engineParam == querypb.EngineType_default {
 		engineParam = g.defaultEngine
 	}
 
+	var qry promql.Query
 	switch engineParam {
 	case querypb.EngineType_prometheus:
-		engine = g.engineFactory.GetPrometheusEngine()
+		queryEngine := g.engineFactory.GetPrometheusEngine()
+		qry, err = queryEngine.NewRangeQuery(ctx, queryable, promql.NewPrometheusQueryOpts(false, lookbackDelta), request.Query, startTime, endTime, interval)
+		if err != nil {
+			return err
+		}
 	case querypb.EngineType_thanos:
-		engine = g.engineFactory.GetThanosEngine()
+		thanosEngine := g.engineFactory.GetThanosEngine()
+		plan, err := logicalplan.Unmarshal(request.QueryPlan.GetJson())
+		if err != nil {
```

Comment:
Since the plan is optional, I suggest we fall back to `queryEngine.NewRangeQuery` and use the string representation.

Reply:
Done